### PR TITLE
Fixed #27830 -- Used distutils.version.LooseVersion for version parsing.

### PR DESCRIPTION
**Pull request by ChillarAnand**
https://code.djangoproject.com/ticket/27830

### CODE REVIEW

**Comment by timgraham** [comment]
Why are tests removed? I'd expect more tests to be added to show what cases this fixes.

**Comment by ChillarAnand** [comment]
@timgraham Added relevant test cases.

File: django/db/backends/postgresql/base.py [hunk 1]
```diff
@@ -23,12 +24,12 @@
     raise ImproperlyConfigured("Error loading psycopg2 module: %s" % e)
 
 
-def psycopg2_version():
+def psycopg2_version_tuple():
```
Comment (timgraham) [inline django/db/backends/postgresql/base.py]: I think this renaming can be omitted.

File: django/contrib/gis/db/backends/postgis/operations.py [hunk 2]
```diff
@@ -353,18 +353,12 @@ def postgis_version_tuple(self):
...
-        m = self.version_regex.match(version)
-
-        if m:
-            major = int(m.group('major'))
-            minor1 = int(m.group('minor1'))
-            minor2 = int(m.group('minor2'))
-        else:
+        version_tuple = get_version_tuple(version)
+        if len(version_tuple) != 3:
             raise Exception('Could not parse PostGIS version string: %s' % version)
```
Comment (timgraham) [inline django/contrib/gis/db/backends/postgis/operations.py]: I'm not sure if this is needed. If keeping it, perhaps a test could be added.

File: django/contrib/gis/db/backends/spatialite/operations.py [hunk 3]
```diff
@@ -189,16 +190,11 @@ def spatialite_version_tuple(self):
...
-
-        m = self.version_regex.match(version)
-        if m:
-            major = int(m.group('major'))
-            minor1 = int(m.group('minor1'))
-            minor2 = int(m.group('minor2'))
-        else:
+        version_tuple = get_version_tuple(version)
+        if len(version_tuple) != 3:
             raise Exception('Could not parse SpatiaLite version string: %s' % version)
```
Comment (timgraham) [inline django/contrib/gis/db/backends/spatialite/operations.py]: I'm not sure if this is needed. If keeping it, perhaps a test could be added.

File: django/contrib/gis/geos/__init__.py [hunk 4]
```diff
@@ -9,7 +9,7 @@
 from .factory import fromfile, fromstr  # NOQA
 from .geometry import GEOSGeometry, hex_regex, wkt_regex  # NOQA
 from .io import WKBReader, WKBWriter, WKTReader, WKTWriter  # NOQA
-from .libgeos import geos_version, geos_version_info  # NOQA
+from .libgeos import geos_version, geos_version_tuple  # NOQA
```
Comment (timgraham) [inline django/contrib/gis/geos/__init__.py]: Not sure if this change is needed -- doesn't seem like Django is importing from here?

File: django/contrib/gis/db/backends/spatialite/operations.py [hunk 5]
```diff
@@ -189,16 +190,11 @@ def spatialite_version_tuple(self):
         minor, subminor).
         """
         version = self.spatialite_version()
-
-        m = self.version_regex.match(version)
```
Comment (timgraham) [inline django/contrib/gis/db/backends/spatialite/operations.py]: Remove unused `version_regex`?

File: django/contrib/gis/geos/libgeos.py [hunk 6]
```diff
@@ -199,3 +200,11 @@ def geos_version_info():
         raise GEOSException('Could not parse version info string "%s"' % ver)
     return {key: m.group(key) for key in (
         'version', 'release_candidate', 'capi_version', 'major', 'minor', 'subminor')}
+
+
+def geos_version_tuple():
+    """
+    Return the GEOS version as a tuple (major, minor, subminor).
+    """
+    version = geos_version_info()['version']
```
Comment (timgraham) [inline django/contrib/gis/geos/libgeos.py]: `get_version_tuple(geos_version_info()['version'])` is okay

File: django/utils/version.py [hunk 7]
```diff
@@ -77,3 +78,14 @@ def get_git_changeset():
     except ValueError:
         return None
     return timestamp.strftime('%Y%m%d%H%M%S')
+
+
+def get_version_tuple(version):
+    """Return tuple version (X, Y, Z) from VERSION if present."""
```
Comment (timgraham) [inline django/utils/version.py]: "from VERSION if present" seems inaccurate

File: tests/backends/tests.py [hunk 8]
```diff
@@ -321,15 +321,14 @@ def test_lookup_cast(self):
...
-    def test_correct_extraction_psycopg2_version(self):
-        from django.db.backends.postgresql.base import psycopg2_version
-        version_path = 'django.db.backends.postgresql.base.Database.__version__'
+    def test_psycopg2_version_extraction_and_comparision(self):
+        from django.db.backends.postgresql.base import psycopg2_version_tuple
 
-        with mock.patch(version_path, '2.6.9'):
-            self.assertEqual(psycopg2_version(), (2, 6, 9))
+        with mock.patch('psycopg2.__version__', '4.2.1 (dt dec pq3 ext lo64)'):
+            self.assertTrue(psycopg2_version_tuple(), (4, 2, 1))
```
Comment (timgraham) [inline tests/backends/tests.py]: Switching to `assertTrue()` seems incorrect.

File: tests/gis_tests/tests.py [hunk 9]
```diff
@@ -68,6 +68,12 @@ def test_version_dev_tuple(self):
         actual = ops.postgis_version_tuple()
         self.assertEqual(expect, actual)
 
+    def test_version_git_tuple(self):
```
Comment (timgraham) [inline tests/gis_tests/tests.py]: git?

File: django/utils/version.py (hunk 7 above)
Comment (ChillarAnand) [inline django/utils/version.py]: Does `Return available versions (major, minor, patch) from VERSION` sound accurate?

Comment (timgraham) [inline django/utils/version.py]: Why is VERSION capitalized? Maybe something like:

Return a tuple of version numbers (e.g. (1, 2, 3)) from the version string (e.g. '1.2.3').

Comment (ChillarAnand) [inline django/utils/version.py]: `VERSION` is capitalized in all other doc strings from this file. So, I thought it should be.

**Pull request merged**